### ISSUE TITLE
Can GeneratorParams take on their actual values in their constructors?

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#603]
This would allow a Param/ImageParam ctor to properly rely on the value of a GeneratorParam.

abadams, from email conversation:

"For a generator called T, I think the current sequence of events is:

in RegisterGenerator<T>::TFactory::create():
  GeneratorBase constructor
  Generator<T> constructor
  T member constructors (i.e. GeneratorParam constructors)
  T constructor
  GeneratorBase::set_generator_param_values

later:
  T::build()

What if instead the Generator<T> constructor took the GeneratorParamValues, and then GeneratorParams went and retrieved their values in their constructors (via looking up the Generator<T> in the object registry)?"

**Comment by steven-johnson** [comment]
So I thought about this, but I don't think it will work: 

(1) to make use of this, a Generator would have to have a ctor that took GeneratorParamValues (either by adding an explicit ctor or by explicitly inheriting a ctor via "using GeneratorBase::GeneratorBase;"); I don't think we can quietly make it work for existing Generator classes. 

(2) even if we could, we'd be treading on dangerous ground: the Generator ctor would be running before the T ctor; if we called set_generator_param_values() from the Generator ctor, the GeneratorParams in T wouldn't be initialized yet, and would get stomped on when T ctor actually runs subsequently. (We could probably take hacky steps to prevent that, but eww.)

There's probably a way to improve this, but I'm not sure this it it.

**Comment by abadams** [comment]
Yeah, I went through roughly the same thought process regarding the ctor. I
concluded that the GeneratorParams need to set their own values in their
constructors by going and finding the GeneratorParamValues through some
back channel.

One option is to register the GeneratorParamValues as relating to the
memory area that the Generator is in before constructing the Generator
(which would require a placement new). The GeneratorParams would then check
if there was a GeneratorParamValues instance registered for a region
including their memory address in their constructor. I think it's possible,
but this was gross enough that I didn't proceed.

**Comment by steven-johnson** [comment]
As an FYI, I've found a place where this is an actual pain point rather than a suboptimal design:

A common pattern is to customize the type of an ImageParam, something like:

    class Foo : Generator<Foo> {
        GeneratorParam<Type> input_type_{"input_type", UInt(8)};
        ImageParam input_{UInt(8), 3, "input"};  // UInt(8) is a placeholder
        Func build() override {
            input_ = ImageParam(input_type_, input_.dimensions(), input_.name());
        }
    }

If we introspect this Generator in order to build argument metadata (see issue 664), we can't rely on the Params/ImageParams being correct unless build() has been called. (e.g., if we instantiated this with input_type=float32, we'd have incorrect metadata.)

To solve this specific problem, we could probably just ensure that we call the build() method before introspection.... pragmatic but kinda eww.

As a more involved workaround, we could add a new method to override that is intended to do this sort of FixUpTheParams() work (but less than all of build())... not sure that gains us anything than a more confusing set of conventions, though (especially since I'm not sure if we can easily enforce a restriction against mutating Params in the build() method).

**Comment by abadams** [comment]
There is a configure() phase that handles this.
